### ISSUE TITLE
Using PropertyName value as enum value

### ISSUE THREAD

**Issue by wikes82** [firebase/firebase-android-sdk#239]
## What feature would you like to see?
saving PropertyName annotation value for enum, instead of enum name

## How would you use it?

```
public enum MyEnum {
  @PropertyName("one")
   ONE("one"),
 @PropertyName("secondOne")
   SECOND_ONE("secondOne")
}
```
so, instead saving it into Firestore as ONE or SECOND_ONE,  we use PropertyName value of one or secondOne.

We generate our POJO using jsonschema2pojo, and it capitalized all enum names, see : https://github.com/joelittlejohn/jsonschema2pojo/wiki/Reference#enum

We use custom annotator to add PropertyName annotation into our POJO but can't change the way jsonschema2pojo behave on naming the enum name without modifying their source codes.

**Comment by wilhuff** [comment]
This seems like a pretty good idea!

Unfortunately, we're completely booked on other work right now so if you're motivated to make this happen your best path forward would be to create a pull request that implements it.

The code that does the POJO mapping is in [CustomClassMapper](https://github.com/firebase/firebase-android-sdk/blob/master/firebase-firestore/src/main/java/com/google/firebase/firestore/util/CustomClassMapper.java). In particular you'll see that when mapping:

  * from an object we're unconditionally just taking the [name of the enumeration member](https://github.com/firebase/firebase-android-sdk/blob/26e848e07812e86b83773c7376b1787023ab41e4/firebase-firestore/src/main/java/com/google/firebase/firestore/util/CustomClassMapper.java#L158)
  * to an object we're unconditionally [calling Enum.valueOf with the name](https://github.com/firebase/firebase-android-sdk/blob/26e848e07812e86b83773c7376b1787023ab41e4/firebase-firestore/src/main/java/com/google/firebase/firestore/util/CustomClassMapper.java#L322).

In the forward direction it should be possible to either use [annotatedName](https://github.com/firebase/firebase-android-sdk/blob/26e848e07812e86b83773c7376b1787023ab41e4/firebase-firestore/src/main/java/com/google/firebase/firestore/util/CustomClassMapper.java#L930) or do something similar.

The reverse direction is trickier but the straightforward thing that just iterates over the enumeration members to find a match is probably OK as a fallback from `Enum.valueOf`.

I'd start by forking the repo, and then following along with our [getting started instructions](https://github.com/firebase/firebase-android-sdk/#getting-started). Make sure `./gradlew :firebase-firestore:check` passes before making changes.

Then, add a test to [MapperTest](https://github.com/firebase/firebase-android-sdk/blob/master/firebase-firestore/src/test/java/com/google/firebase/firestore/util/MapperTest.java) and set breakpoints/print statements near what I described above to verify my recollection :-).

**Comment by mikelehen** [comment]
Looks like the PR was merged and will be included in an upcoming release.  Thanks very much for raising this to our attention and contributing the code!  Much appreciated. 😀 

Just to set expectations, this will probably miss the next Firebase Android release, but be included in the one after (in approximately 3-4 weeks).

**Comment by iamMehedi** [comment]
has this been released?

**Comment by mikelehen** [comment]
@iamMehedi I believe it's included in 18.2.0 which came out Wednesday.
